integrations/integrations.go

```diff
@@ -0,0 +1,231 @@
+
+type fakeTraceWriter struct{}
+
+func (f *fakeTraceWriter) Write([]byte) (int, error) {
+	return 0, nil
+}
+
+func (f *fakeTraceWriter) Close() error {
+	return nil
+}
+
+// See https://github.com/cosmos/cosmos-sdk/blob/7fb26685cd68a6c1d199dc270c80f49f2bfe7ace/server/start.go#L624
+func startApp(
+	env *environment.Env,
+	svrCtx *server.Context,
+	appCreator servertypes.AppCreator,
+	opts server.StartCmdOptions,
+) (app servertypes.Application, err error) {
+	db, err := opts.DBOpener(svrCtx.Config.RootDir, server.GetAppDBBackend(svrCtx.Viper))
+	if err != nil {
+		return app, fmt.Errorf("open db: %v", err)
+	}
+
+	// TODO: Check if is testnet and implement `testnetify` function
+
+	app = appCreator(svrCtx.Logger, db, &fakeTraceWriter{}, svrCtx.Viper)
+	env.Defer(func() {
+		if localErr := app.Close(); localErr != nil {
+			svrCtx.Logger.Error(localErr.Error())
+		}
+	})
+
+	return app, nil
+}
+
+// See https://github.com/cosmos/cosmos-sdk/blob/7fb26685cd68a6c1d199dc270c80f49f2bfe7ace/server/start.go#L307
+// We modify this function to start a Monomer node in-process instead of a Comet node.
+func startInProcess(
+	env *environment.Env,
+	svrCtx *server.Context,
+	clientCtx *client.Context,
+	app servertypes.Application,
+	opts server.StartCmdOptions,
+) error {
+	g, ctx := getCtx(svrCtx, true)
+
+	svrCtx.Logger.Info("Starting Monomer node in-process")
+	wrappedApp := &WrappedApplication{app}
+	_, err := startMonomerNode(wrappedApp, env, svrCtx)
+	if err != nil {
+		return fmt.Errorf("start Monomer node: %v", err)
+	}
+
+	if opts.PostSetup != nil {
+		if err := opts.PostSetup(svrCtx, *clientCtx, ctx, g); err != nil {
+			return fmt.Errorf("run post setup: %v", err)
+		}
+	}
+
+	return g.Wait()
+}
+
+// Starts the Monomer node in-process in place of the Comet node. The
+// `MonomerGenesisPath` flag must be set in viper before invoking this.
+func startMonomerNode(wrappedApp *WrappedApplication, env *environment.Env, svrCtx *server.Context) (*node.Node, error) {
+	engineWS, err := net.Listen("tcp", viper.GetString(monomerEngineWSFlag))
+	if err != nil {
+		return nil, fmt.Errorf("create Engine websocket listener: %v", err)
+	}
+	env.DeferErr("close engine ws", engineWS.Close)
+
+	cmtListenAddr := svrCtx.Config.RPC.ListenAddress
+	cmtListenAddr = strings.TrimPrefix(cmtListenAddr, "tcp://")
+	cometListener, err := net.Listen("tcp", cmtListenAddr)
+	if err != nil {
+		return nil, fmt.Errorf("create CometBFT listener: %v", err)
+	}
+	env.DeferErr("close comet listener", cometListener.Close)
+
+	appdb, err := cometconfig.DefaultDBProvider(&cometconfig.DBContext{
+		ID:     "app",
+		Config: svrCtx.Config,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("create app db: %v", err)
+	}
+	env.DeferErr("close app db", appdb.Close)
+	blockdb := dbm.NewMemDB()
+	env.DeferErr("close block db", blockdb.Close)
+	txdb := cometdb.NewMemDB()
+	env.DeferErr("close tx db", txdb.Close)
+	mempooldb := dbm.NewMemDB()
+	env.DeferErr("close mempool db", mempooldb.Close)
```

Comment:
Apologies, I think I was incorrect before. We can use the provided `DBOpener` for the application as used [here](https://github.com/cosmos/cosmos-sdk/blob/929803d6ccbbb25b05dd605ca50f3ea06905e359/server/start.go#L631).

For the Monomer dbs (block db, tx db, mempool db), we need to get creative. Since Monomer is replacing CometBFT from the SDK's perspective, we should use Comet's configuration to figure out what database we should use. The `DefaultDBProvider` as used [here](https://github.com/cosmos/cosmos-sdk/blob/929803d6ccbbb25b05dd605ca50f3ea06905e359/server/start.go#L782) (similar to your usage for the app db I believe) should work for us.

In the tests, we can configure the `DBOpener` (app db) and `DefaultDBProvider` (block db, tx db, mempool db) to use in-memory databases.

Does this hopefully make things clearer? Apologies again for the confusion here.

Reply:
Yes this clears things up. Thank you! I'll make the change

Reply:
I noticed that `DefaultDBProvider` returns a `cometbft-db.DB` object, but our `node.New` expects `blockdb` and `mempooldb` to be of type `cosmos-db.DB` ...